### PR TITLE
Search: Improve reliability of search class initialization

### LINKED ISSUE #22801: Search: Be more defensive in logic around debug bar panel
### Impacted plugin

Jetpack

### Steps to Reproduce

We have not been able to reproduce. But, choosing to be defensive here will save us future issues in cases where users are interacting in ways that we don't expect.

### A clear and concise description of what you expected to happen.

See report at https://github.com/Automattic/vip-go-mu-plugins/issues/2891

### What actually happened

See report at https://github.com/Automattic/vip-go-mu-plugins/issues/2891

### Other information

p1644522524900439-slack-CDD9LQRSN

### Operating System

_No response_

### OS Version

_No response_

### Browser

Other / Not applicable

### Browser Version(s)

_No response_

### PR DESCRIPTION
**Pull request by jsnmoon**
Fixes #22801.

#### Changes proposed in this Pull Request:

* Change `Classic_Search::instance` and `Instant_Search::instance` functions to accept an optional blog ID parameter.
* Make `Classic_Search::initialize` and `Instant_Search::initialize` to be aliases of the `instance` function.
* Skip rendering the Jetpack Search debug bar if the search instance has not been properly initialized. 

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Apply this change to your site **without** a Search subscription. 
(Note: Block-based theme support is still in development and is not included in this PR.)
* Purchase a Jetpack Search subscription for the site.
* Ensure that the auto-configuration works as expected.
* Ensure that Instant Search and Customberg works as expected.

### CODE REVIEW

File: projects/packages/search/src/classic-search/class-classic-search.php [hunk 1]
```diff
@@ -95,29 +95,26 @@ protected function __construct() {
...
-	/**
-	 * Instantiate and initialize a singleton instance of the class.
+	 * Returns a class singleton. Initializes with first-time setup if given a blog ID parameter.
 	 *
 	 * @param string $blog_id Blog id.
 	 * @return static The class singleton.
 	 */
-	public static function initialize( $blog_id ) {
-		if ( ! isset( self::$instance ) ) {
+	public static function instance( $blog_id = null ) {
```
Comment (kangzj) [inline projects/packages/search/src/classic-search/class-classic-search.php:103]: Could we set `blog_id` to the current blog id inside the initializing process as a fallback? Having `$blog_id` as optional parameter is compatible with the old versions however would still possibly return a un-initialized instance if not used properly.

File: projects/plugins/jetpack/modules/search.php [hunk 2]
```diff
@@ -40,6 +40,7 @@ protected function __construct() {
 	 * Return the instance of the new class.
 	 */
 	public static function instance() {
-		return Classic_Search::initialize( get_current_blog_id() );
+		// Explicitly provide the blog ID, just in case.
+		return Classic_Search::instance( Jetpack_Options::get_option( 'id' ) );
```
Comment (kangzj) [inline projects/plugins/jetpack/modules/search.php:44]: I guess the code might not be needed on WPCOM at all, but probably better be compatible as well to get the blog_id?

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 3]
```diff
@@ -750,15 +750,7 @@ public function update_data( $request ) {
...
-								$instance = Automattic\Jetpack\Search\Instant_Search::instance();
-
-								// Perform initialization. This should have already been executed in modules/search.php.
-								// In other words: This is a failsafe.
-								if ( ! $instance ) {
-									Automattic\Jetpack\Search\Jetpack_Initializer::initialize();
-									$instance = Automattic\Jetpack\Search\Instant_Search::instance();
-								}
-
+								$instance = Automattic\Jetpack\Search\Instant_Search::instance( Jetpack_Options::get_option( 'id' ) );
```
Comment (kangzj) [inline projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php:753]: I guess the code might not be needed on WPCOM at all, but probably better be compatible as well to get the blog_id?

**Review by kangzj — DISMISSED** [review]
It works as described. Added a couple of comments. 👍 
(sorry for accidentally clicking the approve button)

**Comment by kraftbj** [comment]
If the method name is going to change, a note that I'll need to update https://github.com/alleyinteractive/es-wp-query/pull/94

Are we planning on keeping that stable moving forward?

File: projects/packages/search/src/classic-search/class-classic-search.php (hunk 1 above)
Comment (jsnmoon) [reply projects/packages/search/src/classic-search/class-classic-search.php:103]: Good point, I'll implement this change.

File: projects/plugins/jetpack/modules/search.php (hunk 2 above)
Comment (jsnmoon) [reply projects/plugins/jetpack/modules/search.php:44]: While we don't intend for this shim class to be synced or used in WPCOM, I suppose it won't hurt.

File: projects/packages/search/src/classic-search/class-classic-search.php [hunk 4]
```diff
@@ -101,7 +101,10 @@ protected function __construct() {
 	 * @return static The class singleton.
 	 */
 	public static function instance( $blog_id = null ) {
-		if ( null !== $blog_id && ! isset( self::$instance ) ) {
+		if ( ! isset( self::$instance ) ) {
+			if ( null === $blog_id ) {
+				$blog_id = get_current_blog_id();
```
Comment (kangzj) [inline projects/packages/search/src/classic-search/class-classic-search.php:106]: sorry to be a pain. Just a bit confused by how we get `blog_id`. AFAIK, `get_current_blog_id` right? If we need this again:
```
defined( 'IS_WPCOM' ) && constant( 'IS_WPCOM' ) ) ? get_current_blog_id() : Jetpack_Options::get_option( 'id' )
```

Could we extract it into `Helper` class or somewhere?

Comment (jsnmoon) [reply projects/packages/search/src/classic-search/class-classic-search.php]: Fixed!

File: projects/packages/search/src/classic-search/class-classic-search.php [hunk 5]
```diff
@@ -95,29 +95,29 @@ protected function __construct() {
...
 		}
-
 		return self::$instance;
 	}
 
+	/**
+	 * Alias of the instance function.
+	 */
+	public static function initialize() {
+		return call_user_func_array( array( 'self', 'instance' ), func_get_args() );
```
Comment (kangzj) [inline projects/packages/search/src/classic-search/class-classic-search.php:118]:
```suggestion
		return call_user_func_array( array( self::class, 'instance' ), func_get_args() );
```

Comment (kangzj) [inline projects/packages/search/src/classic-search/class-classic-search.php:118]: This blows my mind, but works 👍  I thought `self` is a bit ambiguous here, but it works without any issues.

File: projects/packages/search/src/class-helper.php [hunk 6]
```diff
@@ -923,4 +923,19 @@ public static function get_active_plugins() {
...
+	 * @return int Blog ID.
+	 */
+	public static function get_wpcom_site_id() {
+		// Returns local blog ID for a multi-site network.
+		if ( defined( 'IS_WPCOM' ) && constant( 'IS_WPCOM' ) ) {
+			return get_current_blog_id();
+		}
+
+		// Returns cache site ID.
+		return Jetpack_Options::get_option( 'id' );
```
Comment (kangzj) [inline projects/packages/search/src/class-helper.php:939]:
```suggestion
		return \Jetpack_Options::get_option( 'id' );
```

File: projects/packages/search/src/class-helper.php [hunk 7]
```diff
@@ -923,4 +923,19 @@ public static function get_active_plugins() {
...
+
+	/**
+	 * Get the current site's WordPress.com ID.
+	 *
+	 * @return int Blog ID.
+	 */
+	public static function get_wpcom_site_id() {
+		// Returns local blog ID for a multi-site network.
+		if ( defined( 'IS_WPCOM' ) && constant( 'IS_WPCOM' ) ) {
+			return get_current_blog_id();
```
Comment (kangzj) [inline projects/packages/search/src/class-helper.php:935]:
```suggestion
			return \get_current_blog_id();
```

**Review by kangzj — DISMISSED** [review]
It works well and changes look reasonable to me 👍

**Review by jeherve — APPROVED** [review]
This seems to test well for me. 🚢

**Pull request merged**